Api Gateway Dual GW Arch model documentation update (#2500)
hashicorp/web-unified-docs: content/consul/v2.0.x/content/docs/north-south/api-gateway/k8s/index.mdx

Reviewer: ```suggestion
API gateways enable external network clients to access applications and services running in a Consul datacenter. API gateways can also forward requests from clients to specific destinations based on path or request protocol. Systems that access services in the mesh may be internal or external to your organizational network. _North-south traffic_ is a common term to describe this type of network traffic.
```

This article predates the "Consul API gateway" concept being different to the "Standard API gateway", and uses the term "Consul API gateway" while actually referring to the Standard one. Unless specifically referring to the recently introduced "Consul API gateway", please convert all "Consul API gateway" terms into "API gateway"

Reviewer 2: Accepted the suggestion.

```yaml
apiVersion: v1
items:
- apiVersion: gateway.networking.k8s.io/v1
  kind: Gateway
  metadata:
    labels:
      component: api-gateway
    name: api-gateway
    namespace: consul
  spec:
    gatewayClassName: consul
    listeners:
    - allowedRoutes:
        namespaces:
          from: All
      name: http
      port: 80
      protocol: HTTP
    - allowedRoutes:
        namespaces:
          from: All
      name: tcp-8090
      port: 8090
      protocol: TCP
```

</CodeBlockConfig>

</Tab>

<Tab heading="Consul API Gateway example">

This is an example manifest of Consul API Gateway.

<CodeBlockConfig filename="gateway.yaml">

```yaml
apiVersion: v1
items:
- apiVersion: consul.hashicorp.com/v1beta1
  kind: Gateway
  metadata:
    labels:
      component: api-gateway-consul
    name: api-gateway-consul
    namespace: consul
  spec:
    gatewayClassName: consul-custom-class
    listeners:
    - allowedRoutes:
        namespaces:
          from: All
      name: http
      port: 80
      protocol: HTTP
    - allowedRoutes:
        namespaces:
          from: All
      name: tcp-8090
      port: 8090
      protocol: TCP
```

</CodeBlockConfig>

</Tab>

> </Tabs>

Reviewer 3: I think this section might be unnecessary. In my [comment above](https://github.com/hashicorp/web-unified-docs/pull/2500/changes#r3597509699) i suggest a new structure for the page. I think the examples could go in the two H3 in that section.

This avoids:
* The very long header title
* Having a bullet point list made of only one item

To avoid the page being too long, you can use the accordion to make it foldable instead of using the tabs.

Reviewer 3: Also the condition:

> Your Kubernetes platform manages the Gateway CRDs and only supports stable API versions, and your applications use TCPRoute. In this case, use Consul API Gateway because the Standard API Gateway does not support TCPRoute

Could you please provide examples of platforms that would require this?

Reviewer 2: * Gave details about Standard API gateway and Consul API gateway in introduction itself.
*Could not find any collapsible tags to show the examples into, used Tabs in a section ##Example manifests

API gateway implements the Kubernetes Gateway API specification. For more information, refer to [the official Kubernetes documentation](https://gateway-api.sigs.k8s.io/)

There are two API gateway controllers:

- Standard API gateway, API group `gateway.networking.k8s.io`
- Consul API gateway, API group `consul.hashicorp.com`

### Standard API gateway

We recommend the Standard API Gateway, developed under the `gateway.networking.k8s.io` API group and accepted by the Kubernetes community.

Use the Standard API Gateway in all cases **except** the following:

- Your Kubernetes platform manages the Gateway CRDs, supports only stable API versions (`v1`), and your applications use `TCPRoute`. In this case, use the Consul API Gateway instead, as the Standard API Gateway does not support `TCPRoute` under these constraints.

> 

Reviewer 3: ```suggestion

The Standard API gateway is the default implementation. Developed under the `gateway.networking.k8s.io` is already used in previous Consul releases, tested, and accepted by the Kubernetes community.

We recommend the Standard API gateway for every new installation except when your Kubernetes platform manages the Gateway CRDs, supports only stable API versions (`v1`), and your applications use `TCPRoute`. 

In this case, use the Consul API Gateway instead, as the Standard API Gateway does not support `TCPRoute` under these constraints.

```

### Consul API gateway

Consul API Gateway is a Consul-specific API gateway developed under the `consul.hashicorp.com` API group. Use the Consul API gateway for platforms where gateway CRDs are managed externally and are constrained to gateway API versions earlier than `1.6.0`, supporting only the stable `v1` API version.

Use the Consul API gateway in the following cases:

1. The platform manages the gateway CRDs, and the CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.

> 1. An external application or user installed the gateway CRDs. The CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.

Reviewer 3: ```suggestion
Consul API Gateway is a Consul-specific API gateway introduced with Consul 2.0.0. Developed under the `consul.hashicorp.com` API group. 

The Consul API gateway is useful for the following use cases:
1. The platform manages the gateway CRDs, and the CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.
1. An external application or user installed the gateway CRDs. The CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.

For all cases we recommend the Standard API gateway.
```

Reviewer 3: We give different information in different sections:

Inside [## Standard API gw](https://github.com/hashicorp/web-unified-docs/pull/2500/changes#r3683203364) we say:

> We recommend the Standard API gateway for every new installation except when your Kubernetes platform manages the Gateway CRDs, supports only stable API versions (`v1`), and your applications use `TCPRoute`. 

Meaning that:

(k8s manages CDRs) **AND** (k8s supports only v1 API) **AND** (application uses TCPRoute)

Inside [## Consul API gw](https://github.com/hashicorp/web-unified-docs/pull/2500/changes#r3683238463) we say:

> Use the Consul API gateway in the following cases:
> 1. The platform manages the gateway CRDs, and the CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.
> 1. An external application or user installed the gateway CRDs. The CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.

Meaning that:
((k8s manages CDRs) **OR** (CDRs were installed "externally")) **AND** (k8s supports only v1 API) **AND** (Gateway API `< 1.6.0`) 

I do not know if Gateway API `< 1.6.0` and TCPRoute are related but we should have the same conditions in both sections otherwise is not clear which condition is required for Consul API gw.

## Introduction

API gateways enable external network clients to access applications and services running in a Consul datacenter. API gateways can also forward requests from clients to specific destinations based on path or request protocol. Systems that access services in the mesh may be internal or external to your organizational network. _North-south traffic_ is a common term to describe this type of network traffic.

API gateway implements the Kubernetes Gateway API specification. For more information, refer to [the official Kubernetes documentation](https://gateway-api.sigs.k8s.io/)

> Starting Consul 2.0, a new API gateway controller is introduced under the API group `consul.hashicorp.com`. The API gateway deployed under it is called a Consul API gateway. The old(existing) API gateway under the API group `gateway.networking.k8s.io` is called the Standard API gateway.

Reviewer 4: ```suggestion
There are two API gateway controllers:

- Standard API gateway, API group `gateway.networking.k8s.io`
- Consul API gateway, API group `consul.hashicorp.com`

```

### Standard API gateway

> Standard API Gateway is the recommended API gateway, developed under the `gateway.networking.k8s.io` API group and accepted by the Kubernetes community.

Reviewer 4: ```suggestion
We recommend the Standard API Gateway, developed under the `gateway.networking.k8s.io` API group and accepted by the Kubernetes community.
```

Use the Standard API Gateway in all cases **except** the following:

- Your Kubernetes platform manages the Gateway CRDs, supports only stable API versions (`v1`), and your applications use `TCPRoute`. In this case, use the Consul API Gateway instead, as the Standard API Gateway does not support `TCPRoute` under these constraints.

<Note>

> Starting from August 2026, the Standard API gateway supports TCPRoute. We recommend using the Standard API gateway for all use cases, including TCPRoute.

Reviewer 4: ```suggestion
The Standard API gateway supports TCPRoute. We recommend using the Standard API gateway for all use cases, including TCPRoute.
```

Reviewer 2: Not taking up the suggestion as the support fro TCProute in Standard API gateway will be from Aug 2026. Hence mentioning that time is necessary.

Reviewer 4: We do not publish content for future features. We expect the current docs to reflect the features in the latest release.

If the TCProute in Standard API gateway feature isn't going to be available until after a specific date in August, then this PR should not be published until after that date OR the TCP route content should be in a separate PR.

Reviewer 2: Got it. Removed the whole sentence to avoid confusions.

</Note>

### Consul API gateway

> Consul API Gateway is a Consul-specific API gateway developed under the `consul.hashicorp.com` API group. It is designed for platforms where Gateway CRDs are managed externally and are constrained to Gateway API versions earlier than `1.6.0`, supporting only the stable `v1` API version.

Reviewer 4: ```suggestion
Consul API Gateway is a Consul-specific API gateway developed under the `consul.hashicorp.com` API group. Use the Consul API gateway for platforms where gateway CRDs are managed externally and are constrained to gateway API versions earlier than `1.6.0`, supporting only the stable `v1` API version.
```

> Use the Consul API Gateway in the following cases:

Reviewer 4: ```suggestion
Use the Consul API gateway in the following cases:
```

> 1. Gateway CRDs are managed by the platform and support only the stable `v1` API version with Gateway API `< 1.6.0`.

Reviewer 4: ```suggestion
1. The platform manages the gateway CRDs, and the CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.
```

> 2. Gateway CRDs are installed by an external application or user and support only the stable `v1` API version with Gateway API `< 1.6.0`.

Reviewer 4: ```suggestion
1. An external application or user installed the gateway CRDs. The CRDs support only the stable `v1` API version with Gateway API `< 1.6.0`.
```

Consul API gateway implements the Kubernetes Gateway API specification. For more information, refer to [the official Kubernetes documentation](https://gateway-api.sigs.k8s.io/)

> Starting Consul 2.0, A new API Gateway controller is introduced under the API group `consul.hashicorp.com`, the API Gateway deployed under it is called a Consul API Gateway. The old API Gateway which is under the API group `gateway.networking.k8s.io` will be called as Standard API Gateway.

Reviewer: If using Standard API GW is the recommended way, then calling the Consul API GW "new" will confuse users because most people will assume new means better.

Last but not least, this paragraph does not feel correctly placed into the Introduction section. The overall function of API gateways should be transparent to whether you use Standard or Consul ones. Please move this paragraph into the next section, after explaining how to choose which one and then defining how to use the chosen one.

Reviewer 2: I believe giving a hint of new feature in introduction is good as we are already doing. Then the next section of "How to choose the API Gateway between Standard API Gateway and Consul API Gateway" makes sense and the user knows what they are choosing between.

Reviewer 3: I agree with Krastin over the confusion generated by this section.

I think this could be a section in itself, something like:

```
## Available API gateway implementation

Starting Consul 2.0, A new API Gateway controller is introduced under the API group `consul.hashicorp.com`, the API Gateway deployed under it is called a Consul API Gateway. The old API Gateway which is under the API group `gateway.networking.k8s.io` will be called as Standard API Gateway.

### Standard API Gateway

What it is and why it is the recommended choice. The intuitive flow would be "new is always better" so we should explain a bit why this is the recommended choice.

### Consul API Gateway

What it is and why it was introduced.
```

## How to choose the API Gateway between Standard API Gateway and Consul API Gateway?

Standard API Gateway is the recommened gateway in all the cases excpet the following scenario:

> - Kubernetes platform which manages the Gateway CRDs, set to support only stable version and the user has applications using TCPRoute. In this case, user can deploy Consul API Gateway since the standard gateway fails in supporting TCPRoute.

Reviewer: If the following facts are true:
- using Standard API gateway is recommended and the default
- you cannot use TCPRoute with stable versions( of...?) and Standard API gateway

then using "Consul API gateway" is an exceptional use-case, therefore it should be treated as such. And if it depends on using TCPRoute with stable versions (of what? Kubernetes? Consul? Consul-k8s? Consul-dataplane? I am assuming Standard API gateway but it's unclear) then the section defining this exceptional use case shall be titled
 
`## TCPRoutes on stable Standard API gateway`

```suggestion
## TCPRoutes on stable Standard API gateway

The Standard API gateway does not support using TCPRoutes on a Kubernetes cluster that manages the Gateway CRDs, but supports only stable versions (of ?). Therefore, if you need to use TCPRoutes for your API gateway, you must deploy Consul API Gateway instead.
```

Reviewer 2: The reason of introducing the consul Api gateway feature is because Standard Api gateway did not support TCProute.
Hence , in the section How to choose the API gateway between Standard API Gateway and Consul API gateway we are explaining the details.

Consul API gateway implements the Kubernetes Gateway API specification. For more information, refer to [the official Kubernetes documentation](https://gateway-api.sigs.k8s.io/)

> Starting Consul 2.0, we introduced a new API Gateway controller that is under the API group `consul.hashicorp.com`. The old API Gateway which is under the API group `gateway.networking.k8s.io` will be called as Standard API Gateway. The new API Gateway is called as Consul API Gateway.

Reviewer: This content should expand on what is the difference between the gateway controllers other than their API group. Furthermore, this text is better suited to the " Why and when to use a Consul API gateway?" section, expanding on when to use which one - the Consul API gateway or the legacy Consul API gateway.
Also the expression of actions should not be "we did something". Take a look at a random docs page, for example [this one](https://github.com/hashicorp/web-unified-docs/blob/main/content/consul/v1.9.x/content/docs/connect/index.mdx), and notice how the conjugation is built on statements - "thing is this", "thing does that", "thing can do something" and "when you do this, then that happens", "if you want this, do that".

## What is an API gateway?

> API gateways provide an ingress point for service mesh traffic. They add listeners for external traffic and route HTTP requests to services in the mesh. They can also forward requests from clients to specific destinations based on path or request protocol. Systems that access services in the mesh may be internal or external to your organizational network. _North-south traffic_ is a common term to describe this type of network traffic.

Reviewer: This is a duplicate of line 17, therefore the whole section is redundant.

## Why and when to use a Consul API gateway?

> Consul API gateway are useful for:

Reviewer: Either "Consul API gateway is useful for" or "Consul API gateways are useful for"

> 

Reviewer: We should mention the main point of API gateways - to expose service-mesh services to clients outside of the service mesh

- **Load balancing**: API gateways can distribute incoming requests across multiple services, improving performance and reliability.

> - **For kubernetes platform which manage the Gateway CRDs and set to support only stable version and Applications using TCPRoute**: Consul API Gateway supports TCPRoutes, which are used to route traffic to services that use the TCP protocol. 

Reviewer: This is a hyper-specific statement that should not be present in this section. Also, by omitting details, it is also claiming untrue facts:

- The API gateway is not for HTTPRoutes?
- The API gateway is not for unstable version? (of what?)
- Not for Kubernetes that doesn't manage the Gateway CRDs? Then who does?
- TCPRoutes are used to route traffic to services that use the TCP protocol? Then we should use this for HTTP too, since the HTTP protocol is stacked ontop of TCP

Author: This is highlighted for those users using the platforms such as OCP/GKE Autopilot which manage the Gateway CRDs under stable v1. Here, there will be no support for the TCP Route CRD thus the Consul API Gateway unblocks the customer using the TCPRoutes.

> 

Reviewer: The `## Workflow` title is missing

1. [Deploy listeners](/consul/docs/north-south/api-gateway/k8s/listener). Configure one or more listeners to serve as ingress points for requests to services in Consul's service mesh.
1. [Define a route](/consul/docs/north-south/api-gateway/k8s/route) so that the listener can forward the request to the service in the mesh. Routes can be tuned to define complex traffic shaping. For example, you can [Reroute HTTP requests](/consul/docs/north-south/api-gateway/k8s/reroute) to define URL rewrite rules for the services in your Kubernetes cluster.

> ## Consul API gateway reference objects
> 
> Refer to the following table for the API gateway resource details:
> 
> | Resources | Shortname | API Version | Namespaced | Kind |
> | --------- | --------- | ----------- | ---------- | ---- |
> | cgateways | cgtw | `consul.hashicorp.com/v1beta1` | true | Gateway |
> | cgrpcroutes | cgrpcroutes | `consul.hashicorp.com/v1alpha2` | true | GRPCRoute |
> | chttproutes | chttproutes | `consul.hashicorp.com/v1beta1` | true | HTTPRoute |
> | creferencegrants | crefgrant | `consul.hashicorp.com/v1beta1` | true | ReferenceGrant |
> | ctcproutes | ctcproutes | `consul.hashicorp.com/v1alpha2` | true | TCPRoute |
> | ctlsroutes | ctlsroutes | `consul.hashicorp.com/v1alpha2` | true | TLSRoute |
> | cudproutes | cudproutes | `consul.hashicorp.com/v1alpha2` | true | UDPRoute |
> | customgatewayclasses | cgc | `consul.hashicorp.com/v1beta1` | false | CustomGatewayClass |
> | customgatewaypolicies | cgp | `consul.hashicorp.com/v1alpha1` | true | CustomGatewayPolicy |
> 

Reviewer 3: This is referring only to the Consul API gateway correct?

If so apply this suggestion:

```suggestion
## Consul API gateway reference objects

Refer to the following table for the Consul API gateway resource details:

| Resources | Shortname | API Version | Namespaced | Kind |
| --------- | --------- | ----------- | ---------- | ---- |
| cgateways | cgtw | `consul.hashicorp.com/v1beta1` | true | Gateway |
| cgrpcroutes | cgrpcroutes | `consul.hashicorp.com/v1alpha2` | true | GRPCRoute |
| chttproutes | chttproutes | `consul.hashicorp.com/v1beta1` | true | HTTPRoute |
| creferencegrants | crefgrant | `consul.hashicorp.com/v1beta1` | true | ReferenceGrant |
| ctcproutes | ctcproutes | `consul.hashicorp.com/v1alpha2` | true | TCPRoute |
| ctlsroutes | ctlsroutes | `consul.hashicorp.com/v1alpha2` | true | TLSRoute |
| cudproutes | cudproutes | `consul.hashicorp.com/v1alpha2` | true | UDPRoute |
| customgatewayclasses | cgc | `consul.hashicorp.com/v1beta1` | false | CustomGatewayClass |
| customgatewaypolicies | cgp | `consul.hashicorp.com/v1alpha1` | true | CustomGatewayPolicy |

```

## Guidance

> The following resources are available to help you use Consul API gateways on Kubernetes.

Reviewer 3: Is this referring to Consul API gw only?

f not I would rephrase saying something like:

> The following resources are available to help you use API gateways for Consul on Kubernetes.

Reviewer 2: @danielehc  Yes we are referring only the `consul API gateway` related docs below. I think the sentence need not be changed .
